### DISCUSSION TITLE
return RenderFragment from outside razor files

### DISCUSSION THREAD

**Discussion by omuleanu** [dotnet/csharplang#5607] [Language Ideas]
I was trying to return `RenderFragment` from a static class (.cs file) like this: 
```
public static RenderFragment ColSelCaption()
{
    return @<div></div>;
}
```
and I'm getting errors, I can't do this kind of thing neither from component partial/base class, seems that it has to be in `.razor` file for this markup to compile.
1) can I return markup from `.cs` file ?
2) My workaround now is to have a `MyUtils.razor` component which looks like this:
```
@code {
    public static RenderFragment Method1()
    {
        return @<div></div>;
    }
}
```
and I call like in other components using `MyUtils.Method1()`, 
is this workaround ok, could it pose any issues now or in the future ?

**Comment by Liero** [comment]
It's not quite ok for following reasons:

1. It creates component `<MyUtils>` even if you don't use it.
2. If you try to add any EventCallback, it won't compile. E.g.:

    ```razor
    public static RenderFragment FooterTemplate(Action onclick)
    {
        return @<button type="button" @onclick="@onclick">Ok</button>
    }
    ```

    will throw 

    > Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\Shared_MyUtils_razor.g.cs(247,165): error CS0026: Keyword 'this' is not valid in a static property, static method, or static field initializer

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. This would be a feature request for C# language. Moving over to their repo.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one area label.

**Comment by mkArtakMSFT** [comment]
@jaredpar here is a feature request we had talked about a while ago.

**Comment by omuleanu** [comment]
@mkArtakMSFT I could also use a separate Blazor Component instead of each `RenderFragment` function.
Is there any significant performance difference between a `RenderFragment` function (from inside a `.razor` file now) and a separate component ?

**Comment by jaredpar** [comment]
> here is a feature request we had talked about a while ago.

Can you please elaborate? It's unclear reading this issue what the ask is here.

**Comment by Liero** [comment]
@omuleanu: It does has negative performance effect, see https://docs.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-6.0#avoid-thousands-of-component-instances. Whether it is significant or no depends on how many instances you create. But since the issue has been moved to razor repo, it is a bit off topic now

**Comment by mkArtakMSFT** [comment]
@jaredpar the ask is having markup support in C#, so that one can write a C# method (as an example) which will return a markup.

**Reply by omuleanu** [reply]
@mkArtakMSFT from my point of view in doesn't matter if it is a `cs` or `razor` file as long as I can put reusable static methods in there that can return RenderFragment
